src/pip/_internal/resolution/resolvelib/provider.py

```diff
@@ -143,6 +144,22 @@ def get_preference(
             identifier,
         )
 
+    def _get_constraint(self, identifier: str) -> Constraint:
+        with contextlib.suppress(KeyError):
+            return self._constraints[identifier]
+
+        # HACK: Theoratically we should check whether this identifier is a valid
+        # "NAME[EXTRAS]" format, and parse out the name part with packaging or
+        # some regular expression. But since pip's resolver only spits out
+        # three kinds of identifiers: normalized PEP 503 names, normalized names
+        # plus extras, and Requires-Python, we can cheat a bit here.
```

Comment:
I agree, this is a reasonable shortcut to take. Yes, if we ever spit out other types of identifier, this might need to change, but the comment can act as a reminder.